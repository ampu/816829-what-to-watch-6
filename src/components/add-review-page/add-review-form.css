.add-review__form--disabled {
  filter: grayscale(100%);
}

.add-review__form :disabled {
  cursor: default;
}

.add-review__text {
  background-color: rgba(255, 255, 255, 0.24);
}

.add-review__textarea:disabled {
  opacity: 0.5;
}
