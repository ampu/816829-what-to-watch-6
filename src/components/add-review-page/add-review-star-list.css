.rating__input:disabled + .rating__label {
  cursor: default;
}

.rating__input:disabled + .rating__label::after {
  opacity: 0.5;
}
