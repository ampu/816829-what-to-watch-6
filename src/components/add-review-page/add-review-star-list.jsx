import React, {memo} from 'react';
import PropTypes from 'prop-types';

const AddReviewStarList = ({maxRating}) => {
  return Array.from({length: maxRating}, (_, i) => {
    const value = i + 1;
    const id = `star-${value}`;

    return (
      <React.Fragment key={id}>
        <input className="rating__input" id={id} type="radio" name="rating" value={value}/>
        <label className="rating__label" htmlFor={id}>Rating {value}</label>
      </React.Fragment>
    );
  });
};

AddReviewStarList.propTypes = {
  maxRating: PropTypes.number.isRequired,
};

export {AddReviewStarList};

export default memo(AddReviewStarList);
