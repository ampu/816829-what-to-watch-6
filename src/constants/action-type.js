const ActionType = {
  SET_MOVIES_STATUS: `movies/setMoviesStatus`,
  SET_MOVIES: `movies/setMovies`,
  SET_PROMO_STATUS: `movies/setPromoStatus`,
  SET_PROMO_MOVIE: `movies/setPromoMovie`,

  SET_LOGIN_STATUS: `user/setLoginStatus`,
  SET_USER: `user/setUser`,
  SET_MY_LIST_STATUS: `user/setMyListStatus`,
  SET_MY_LIST: `user/setMyList`,

  SET_GENRE: `genres/setGenre`,
};

export default ActionType;
