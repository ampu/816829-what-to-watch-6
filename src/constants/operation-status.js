const OperationStatus = {
  UNSET: `unset`,
  PENDING: `pending`,
  RESOLVED: `resolved`,
  REJECTED: `rejected`,
  UNAUTHORIZED: `unauthorized`,
};

export default OperationStatus;
