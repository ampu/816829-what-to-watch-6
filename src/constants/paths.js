const MainPath = {
  INDEX: `/`,
  SIGN_IN: `/login`,
  MY_LIST: `/mylist`,
  MOVIE: `/films/:id`,
  ADD_REVIEW: `/films/:id/review`,
  PLAYER: `/player/:id`,
  NOT_FOUND: `/not-found`,
};

const MoviePath = {
  MOVIE_OVERVIEW: `/films/:id`,
  MOVIE_DETAILS: `/films/:id/details`,
  MOVIE_REVIEWS: `/films/:id/reviews`,
};

const MOVIE_PATHS = Array.from(new Set([MainPath.MOVIE, ...Object.values(MoviePath)]));

export {
  MainPath,
  MoviePath,
  MOVIE_PATHS,
};
