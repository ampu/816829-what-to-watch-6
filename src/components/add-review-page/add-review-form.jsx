import React, {useState, useCallback} from 'react';
import {Redirect, generatePath} from 'react-router-dom';
import PropTypes from 'prop-types';
import getClassName from 'classnames';

import {usePostReview} from '../../hooks/use-post-review';
import OperationStatus from '../../constants/operation-status';
import {MoviePath} from '../../constants/paths';

import AddReviewStarList from './add-review-star-list';

import {movieType} from '../../typings/movie-type';

import './add-review-form.css';

const MAX_RATING = 10;

const AddReviewForm = ({movie}) => {
  const [text, setText] = useState(``);
  const handleTextChange = useCallback((evt) => {
    setText(evt.target.value);
  }, [setText]);

  const [rating, setRating] = useState(MAX_RATING);
  const handleRatingChange = useCallback((evt) => {
    setRating(+evt.target.value);
  }, [setRating]);

  const [postReviewStatus, postReview] = usePostReview(movie.id);

  const handleFormSubmit = useCallback((evt) => {
    evt.preventDefault();
    postReview({rating, text});
  }, [postReview, rating, text]);

  if (postReviewStatus === OperationStatus.RESOLVED) {
    return <Redirect to={generatePath(MoviePath.MOVIE_REVIEWS, movie)}/>;
  }

  const isDisabled = postReviewStatus === OperationStatus.PENDING;

  const formClassMap = {
    [`add-review__form`]: true,
    [`add-review__form--disabled`]: isDisabled,
  };

  const textClassMap = {
    [`add-review__text`]: true,
    [`add-review__text--disabled`]: isDisabled,
  };

  return (
    <form action="#" className={getClassName(formClassMap)} onSubmit={handleFormSubmit}>
      <div className="rating">
        <div className="rating__stars" value={rating} onChange={handleRatingChange}>
          <AddReviewStarList maxRating={MAX_RATING} isDisabled={isDisabled}/>
        </div>
      </div>

      <div className={getClassName(textClassMap)}>

        <textarea className="add-review__textarea"
          name="review-text" id="review-text" placeholder="Review text" required
          value={text} onChange={handleTextChange} disabled={isDisabled}/>

        <div className="add-review__submit">
          <button className="add-review__btn" type="submit" disabled={isDisabled}>
            {isDisabled ? `Posting` : `Post`}
          </button>
        </div>
      </div>
    </form>
  );
};

AddReviewForm.propTypes = {
  movie: PropTypes.shape({
    id: movieType.id,
  }).isRequired,
};

export default AddReviewForm;
