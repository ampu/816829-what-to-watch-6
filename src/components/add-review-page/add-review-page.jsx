import React from 'react';
import {generatePath, Link, Redirect} from 'react-router-dom';
import {connect} from 'react-redux';

import {MainPath} from '../../constants/paths';
import PosterSize from '../../constants/poster-size';
import OperationStatus from '../../constants/operation-status';
import {selectMovieById, selectMoviesStatus} from '../../store/selectors/movie-selectors';

import Logo from '../logo/logo';
import UserBlock from '../user-block/user-block';
import MoviePoster from '../movie-poster/movie-poster';
import MovieBackground from '../movie-background/movie-background';
import AddReviewForm from './add-review-form';
import SpinnerLoading from '../spinner-loading/spinner-loading';
import Maintenance from '../maintenance/maintenance';
import Container from '../container/container';

import operationStatusType from '../../typings/operation-status-type';
import {movieShape} from '../../typings/movie-type';

const AddReviewPage = ({moviesStatus, movie = {}} = {}) => {

  if (moviesStatus === OperationStatus.PENDING) {
    return (
      <Container isCentered>
        <SpinnerLoading/>
      </Container>
    );
  }

  if (moviesStatus === OperationStatus.REJECTED) {
    return (
      <Container isCentered>
        <Maintenance/>
      </Container>
    );
  }

  if (!movie.id) {
    return <Redirect to={MainPath.NOT_FOUND}/>;
  }

  const {
    title = ``,
    primaryBackgroundStyle,
  } = movie;

  return (
    <section className="movie-card movie-card--full" style={primaryBackgroundStyle}>
      <div className="movie-card__header">
        <MovieBackground movie={movie}/>

        <h1 className="visually-hidden">{title}</h1>

        <header className="page-header">
          <Logo/>

          <nav className="breadcrumbs">
            <ul className="breadcrumbs__list">
              <li className="breadcrumbs__item">
                <Link className="breadcrumbs__link" to={generatePath(MainPath.MOVIE, movie)}>{title}</Link>
              </li>
              <li className="breadcrumbs__item">
                <a className="breadcrumbs__link">Add review</a>
              </li>
            </ul>
          </nav>

          <UserBlock/>
        </header>

        <MoviePoster movie={movie} size={PosterSize.SMALL}/>
      </div>

      <div className="add-review">
        <AddReviewForm movie={movie}/>
      </div>
    </section>
  );
};

AddReviewPage.propTypes = {
  moviesStatus: operationStatusType,
  movie: movieShape,
};

const mapStateToProps = (state, {movieId}) => {
  return {
    moviesStatus: selectMoviesStatus(state),
    movie: selectMovieById(state, movieId),
  };
};

export {AddReviewPage};
export default connect(mapStateToProps)(AddReviewPage);
