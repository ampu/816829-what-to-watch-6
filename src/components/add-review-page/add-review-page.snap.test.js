import React from 'react';
import {Router} from 'react-router-dom';
import {createMemoryHistory} from 'history';
import * as redux from 'react-redux';

import {render} from '@testing-library/react';
import configureStore from 'redux-mock-store';

import {ALL_GENRES} from '../../constants/genre';
import Domain from '../../store/domain';
import OperationStatus from '../../constants/operation-status';

import {AddReviewPage} from './add-review-page';

const INITIAL_STATE = {
  [Domain.MOVIE]: {
    moviesStatus: OperationStatus.UNSET,
    movies: [],
    promoStatus: OperationStatus.UNSET,
    promoMovie: undefined,
  },
  [Domain.USER]: {
    loginStatus: OperationStatus.UNSET,
    user: undefined,
  },
  [Domain.GENRE]: {
    genre: ALL_GENRES,
  },
};

const thunkMock = () => (next) => (action) => {
  if (typeof action === `function`) {
    return Promise.resolve();
  }
  return next(action);
};

const initializeStoreWithState = configureStore([thunkMock]);

describe(`AddReviewPage component`, () => {
  it(`Should render according to the snapshot with pending movies request`, () => {
    const {container} = render(<AddReviewPage moviesStatus={OperationStatus.PENDING}/>);
    expect(container).toMatchSnapshot();
  });

  it(`Should render according to the snapshot with rejected movies request`, () => {
    const {container} = render(<AddReviewPage moviesStatus={OperationStatus.REJECTED}/>);
    expect(container).toMatchSnapshot();
  });

  it(`Should render according to the snapshot with no movie found`, () => {
    const {container} = render((
      <Router history={createMemoryHistory()}>
        <AddReviewPage/>
      </Router>
    ));
    expect(container).toMatchSnapshot();
  });

  it(`Should render according to the snapshot with loaded movie`, () => {
    const movie = {id: `fakeId`};
    const {container} = render((
      <redux.Provider store={initializeStoreWithState(INITIAL_STATE)}>
        <Router history={createMemoryHistory()}>
          <AddReviewPage movie={movie}/>
        </Router>
      </redux.Provider>
    ));
    expect(container).toMatchSnapshot();
  });
});
