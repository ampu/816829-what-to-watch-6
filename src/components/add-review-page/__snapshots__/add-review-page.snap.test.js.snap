// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddReviewPage component Should render according to the snapshot with loaded movie 1`] = `
<div>
  <section
    class="movie-card movie-card--full"
  >
    <div
      class="movie-card__header"
    >
      
      <h1
        class="visually-hidden"
      />
      <header
        class="page-header"
      >
        <div
          class="logo"
        >
          <a
            aria-current="page"
            class="logo__link logo__link--active"
            href="/"
          >
            <span
              class="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              class="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              class="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <nav
          class="breadcrumbs"
        >
          <ul
            class="breadcrumbs__list"
          >
            <li
              class="breadcrumbs__item"
            >
              <a
                class="breadcrumbs__link"
                href="/films/fakeId"
              />
            </li>
            <li
              class="breadcrumbs__item"
            >
              <a
                class="breadcrumbs__link"
              >
                Add review
              </a>
            </li>
          </ul>
        </nav>
        <div
          class="user-block"
        />
      </header>
      
    </div>
    <div
      class="add-review"
    >
      <form
        action="#"
        class="add-review__form"
      >
        <div
          class="rating"
        >
          <div
            class="rating__stars"
            value="10"
          >
            <input
              class="rating__input"
              id="star-1"
              name="rating"
              type="radio"
              value="1"
            />
            <label
              class="rating__label"
              for="star-1"
            >
              Rating 
              1
            </label>
            <input
              class="rating__input"
              id="star-2"
              name="rating"
              type="radio"
              value="2"
            />
            <label
              class="rating__label"
              for="star-2"
            >
              Rating 
              2
            </label>
            <input
              class="rating__input"
              id="star-3"
              name="rating"
              type="radio"
              value="3"
            />
            <label
              class="rating__label"
              for="star-3"
            >
              Rating 
              3
            </label>
            <input
              class="rating__input"
              id="star-4"
              name="rating"
              type="radio"
              value="4"
            />
            <label
              class="rating__label"
              for="star-4"
            >
              Rating 
              4
            </label>
            <input
              class="rating__input"
              id="star-5"
              name="rating"
              type="radio"
              value="5"
            />
            <label
              class="rating__label"
              for="star-5"
            >
              Rating 
              5
            </label>
            <input
              class="rating__input"
              id="star-6"
              name="rating"
              type="radio"
              value="6"
            />
            <label
              class="rating__label"
              for="star-6"
            >
              Rating 
              6
            </label>
            <input
              class="rating__input"
              id="star-7"
              name="rating"
              type="radio"
              value="7"
            />
            <label
              class="rating__label"
              for="star-7"
            >
              Rating 
              7
            </label>
            <input
              class="rating__input"
              id="star-8"
              name="rating"
              type="radio"
              value="8"
            />
            <label
              class="rating__label"
              for="star-8"
            >
              Rating 
              8
            </label>
            <input
              class="rating__input"
              id="star-9"
              name="rating"
              type="radio"
              value="9"
            />
            <label
              class="rating__label"
              for="star-9"
            >
              Rating 
              9
            </label>
            <input
              class="rating__input"
              id="star-10"
              name="rating"
              type="radio"
              value="10"
            />
            <label
              class="rating__label"
              for="star-10"
            >
              Rating 
              10
            </label>
          </div>
        </div>
        <div
          class="add-review__text"
        >
          <textarea
            class="add-review__textarea"
            id="review-text"
            name="review-text"
            placeholder="Review text"
            required=""
          />
          <div
            class="add-review__submit"
          >
            <button
              class="add-review__btn"
              type="submit"
            >
              Post
            </button>
          </div>
        </div>
      </form>
    </div>
  </section>
</div>
`;

exports[`AddReviewPage component Should render according to the snapshot with no movie found 1`] = `<div />`;

exports[`AddReviewPage component Should render according to the snapshot with pending movies request 1`] = `
<div>
  <div
    class="container container--centered"
  >
    <div
      class="spinner-loading"
    >
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
    </div>
  </div>
</div>
`;

exports[`AddReviewPage component Should render according to the snapshot with rejected movies request 1`] = `
<div>
  <div
    class="container container--centered"
  >
    <div
      class="maintenance"
    >
      Maintenance, sorry...
    </div>
  </div>
</div>
`;
